src/server/services/wfs/qgswfsdescribefeaturetype.cpp

```diff
@@ -262,27 +265,86 @@ namespace QgsWfs
         //xsd:element
         QDomElement attElem = doc.createElement( QStringLiteral( "element" )/*xsd:element*/ );
         attElem.setAttribute( QStringLiteral( "name" ), attributeName.replace( ' ', '_' ).replace( cleanTagNameRegExp, QString() ) );
-        QVariant::Type attributeType = fields.at( idx ).type();
+        QVariant::Type attributeType = field.type();
         if ( attributeType == QVariant::Int )
-          attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "integer" ) );
+        {
+          attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "int" ) );
+        }
+        else if ( attributeType == QVariant::UInt )
+        {
+          attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "unsignedInt" ) );
+        }
         else if ( attributeType == QVariant::LongLong )
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "long" ) );
+        }
+        else if ( attributeType == QVariant::ULongLong )
+        {
+          attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "unsignedLong" ) );
+        }
         else if ( attributeType == QVariant::Double )
-          attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "double" ) );
+        {
+          if ( field.length() != 0 && field.precision() == 0 )
+            attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "integer" ) );
+          else
+            attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "decimal" ) );
+        }
         else if ( attributeType == QVariant::Bool )
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "boolean" ) );
+        }
         else if ( attributeType == QVariant::Date )
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "date" ) );
+        }
         else if ( attributeType == QVariant::Time )
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "time" ) );
+        }
         else if ( attributeType == QVariant::DateTime )
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "dateTime" ) );
+        }
         else
+        {
           attElem.setAttribute( QStringLiteral( "type" ), QStringLiteral( "string" ) );
+        }
+
+        const QgsEditorWidgetSetup setup = field.editorWidgetSetup();
+        if ( setup.type() ==  QStringLiteral( "DateTime" ) )
+        {
+          QgsDateTimeFieldFormatter *fieldFormatter = new QgsDateTimeFieldFormatter();
```

Comment:
It seems that there's a memory leak here (`fieldFormatter` is never deleted).

BTW, a pointer is not necessary in this case.